### ISSUE TITLE
[2.2.0] laravel-mix spawning new build in a loop

### ISSUE THREAD

**Issue by vinkla** [tailwindlabs/tailwindcss#4701]
### What version of Tailwind CSS are you using?

2.2.1

### What build tool (or framework if it abstracts the build tool) are you using?

laravel-mix 6.0.23

### What version of Node.js are you using?

v16.3.0

### What browser are you using?

N/A

### What operating system are you using?

macOS

### Reproduction repository

[`vinkla/tailwind-issue-4701`](https://github.com/vinkla/tailwind-issue-4701)

### Describe your issue

Installing the latest version 2.2.0 I'm having an issue with `laravel-mix` spawning a new build in a loop.

https://user-images.githubusercontent.com/499192/122579390-25ce3280-d055-11eb-8533-40095e911fee.mp4

`webpack.mix.js`

```js
mix.postCss('resources/styles/index.css', 'yrgo.css', [
  require('postcss-import'),
  require('tailwindcss/nesting'),
  require('tailwindcss'),
  require('autoprefixer'),
]);
```

`tailwind.config.js`
```js
const defaultTheme = require('tailwindcss/defaultTheme');

module.exports = {
  mode: 'jit',
  purge: ['./public/themes/**/*.php', './resources/**/**/*.{js,css}'],
  darkMode: 'media',
  theme: {
    customForms: (theme) => ({
      default: {
        select: {
          borderColor: theme('colors.gray.400'),
        },
        input: {
          borderColor: theme('colors.gray.400'),
        },
      },
    }),
    extend: {
      fontSize: {
        base: '1.0625rem',
      },
      fontFamily: {
        display: ['Gravur Condensed', ...defaultTheme.fontFamily.sans],
        sans: ['Inter var', ...defaultTheme.fontFamily.sans],
      },
      colors: {
        yrgo: '#e62332',
      },
    },
  },
  plugins: [
    require('@tailwindcss/aspect-ratio'),
    require('@tailwindcss/forms'),
  ],
};

```

**Comment by gregsullivan** [comment]
I'm encountering the same issue. Using `TAILWIND_MODE=watch` has no effect, either.

I think I've isolated the issue to the creation or update of the `mix-manifest.json` file. It seems like this file triggers watch, which then creates it again, forever.

I'm not using Mix with Laravel, and I was able to validate this by adding `.setPublicPath( '../' )` to my `webpack.mix.js` file. Obviously this isn't ideal and isn't a permanent solution, but it seems to confirm the `mix-manifest.json` file is the problem.

Finally, I've also noticed that even when not using JIT, running `watch` will trigger two builds before settling into watching as expected. This is also true when using JIT with the `mix-manifest.json` file being created outside the project folder.

**Comment by tleish** [comment]
I'm not using laravel, but experienced this loop with webpack-dev-server and tailwindcss`mode: 'jit'` after upgrading to v2.2.2.

**Comment by adamwathan** [comment]
Hey! Please provide a reproduction, it is a required field. We would never be able to get through any issues if we had to manually recreate every project from scratch, it's 100x faster if we can just clone something and troubleshoot from there. Please help us out! ❤️

**Comment by gregsullivan** [comment]
@adamwathan I could provide a reproduction of my environment. Do you want a new issue, or within the comments of this issue?

**Comment by adamwathan** [comment]
Here is okay!

**Comment by gregsullivan** [comment]
In the process of creating my reproduction environment, I resolved my issue: I wasn't using the latest version of Node.js on the virtual machine I was using to generate my theme.

However, Mix does now require `TAILWIND_MODE=watch` to work, and it didn't before. For me, that meant changing this line in my package.json:

`"watch": "mix watch",`

to this:

`"watch": "cross-env TAILWIND_MODE=watch mix watch",`

(I recognize the need to use `TAILWIND_MODE=watch` in some circumstances is a known issue while JIT is in preview, so I mention this more for others looking at this issue as opposed to anyone working on Tailwind.)

**Comment by vinkla** [comment]
Reproduction: https://github.com/vinkla/tailwind-issue-4701

**Comment by gregsullivan** [comment]
@vinkla I cloned your repository and changed this line:

`"dev": "mix watch"`

to this:

`"dev": "TAILWIND_MODE=watch mix watch"`

and it worked as expected.

@adamwathan The only other thing I noticed is that now Mix's watch command does two builds when it first runs and then runs normally (with one build per updated file); do you want a separate issue for that?

**Comment by tleish** [comment]
I've discovered the following:

My webpack project loops if I include one of the following patterns in purge:

- `{blog,admin}/app/views/**/*.erb`
- `../*/app/views/**/*.erb`

```
module.exports = {
  mode: 'jit',
  purge: {
    content: [
      './app/packs/**/*.{js,css,scss}',
      '{blog,admin}/app/views/**/*.erb',
      '../*/app/views/**/*.erb',
    ],
  }
```

If I remove these 2 patterns the looping stops.

or

If I set `TAILWIND_MODE=watch`, then the webpack dev server runs without looping, and I can use these 2 patterns (which doesn't feel like a solution to me).

**Comment by tleish** [comment]
Rails and Webpack(er) loop reproduction:

https://github.com/tleish/tailwindcss-issue-4701

**Comment by vinkla** [comment]
> `"dev": "TAILWIND_MODE=watch mix watch"`

@gregsullivan I can confirm that this seems to solve the issue.

> @adamwathan The only other thing I noticed is that now Mix's watch command does two builds when it first runs and then runs normally (with one build per updated file); do you want a separate issue for that?

I can confirm this as well. It is reproducible using the solution above.

**Comment by bradlc** [comment]
Hey all! Tailwind has to register its dependencies with the PostCSS runner, in this case `postcss-loader`. For example if you have `/src/index.html` in your `purge` array then we register that as a file dependency. This tells webpack that it should rebuild the CSS when this file changes. In the case of globs we have to register the "base" directory as a dependency. For example if your glob is `/src/**/*.html` then we have to register the entire `/src` directory as a dependency. The looping issue occurs when your output CSS file is saved into one of the directories that Tailwind has registered as a dependency.

_Note that some runners (e.g. Parcel, Snowpack, Vite) allow us to register the full glob as a dependency which helps to avoid this issue but unfortunately webpack does not support this._

@vinkla In your project you have `./public/themes/**/*.php` in your `purge` config. This means that we register `./public/themes` as a dependency. Because your output CSS file is `./public/themes/undefined/assets/app.css` this causes the loop. When the CSS file is written it triggers a rebuild, which writes the CSS file, which triggers a rebuild...

Our recommendation for avoiding this if you cannot use a different PostCSS runner is to make your `purge` globs more specific and/or rework your directory structure. I am not familiar with WordPlate but for example if you could move your PHP template files into a `templates` directory you could change your purge pattern to `./public/themes/wordplate/templates/**/*.php`.

@tleish The above applies to your project as well, but it's important to note that the base directory of your `{blog,admin}/app/views/**/*.erb` pattern is `.`, because the first segment is dynamic. This means your entire project directory is watched. You can avoid that by splitting it into two separate patterns: `./blog/app/views/**/*.erb` and `./admin/app/views/**/*.erb`

I hope that helps!

**Comment by vinkla** [comment]
Thanks for the thorough explanation Brad 🙌

**Comment by tleish** [comment]
So, to clarify.  If using JIT then all purge GLOB patterns are converted to directories?

To avoid future tickets, you may want to:
1. Update documentation (which provides examples of glob patterns).
2. Fail (or at the least throw a warning) during a build indicating the output css file is inside the the purge path "XYZ" and will cause an infinite loop.

**Comment by tleish** [comment]
BTW, I am curious as to why `TAILWIND_MODE=watch` prevents this issue.  Does write the CSS file to a different location?

**Comment by gregsullivan** [comment]
@tleish This is from the [2.2.0 Changes and deprecations](https://github.com/tailwindlabs/tailwindcss/releases/tag/v2.2.0#changes-and-deprecations) section:

> Set `TAILWIND_MODE=watch` to opt-in to the old dependency tracking system. If the previous default was working for you, this will let you just keep using that system until the build tool you're using has been updated to support the new system.

I was initially confused because webpack is listed as supporting the new dependency system, but converting glob patterns to directories effectively means I can't use Mix due to its reliance on webpack. (It will be more straightforward for me to find another way to run Tailwind than to restructure my project.)

The only outstanding question for me is whether webpack will eventually support registering the full glob as a dependency, in which case using `TAILWIND_MODE=watch` could be viable until that happens.

**Comment by adamwathan** [comment]
@tleish:

> So, to clarify. If using JIT then all purge GLOB patterns are converted to directories?

Not quite, PostCSS supports registering directories as dependencies, and allows you to include a glob optionally but **not all bundlers respect/can support the glob**.

Our Tailwind CLI tool uses the globs exactly as they are, but webpack for example only lets you register directories as dependencies.

We cannot detect what build tool you are using, so it's not possible for us to know when we should warn and when we don't have to. I agree we should update the documentation to explain this gotcha when using webpack though, although it doesn't affect many users because it's much more common to write your output to a different directory than your source files.

> Fail (or at the least throw a warning) during a build indicating the output css file is inside the the purge path "XYZ" and will cause an infinite loop.

This is not possible either because we have no idea where your output file is going to be — that is configured in your build tool and the information is not exposed to us via PostCSS. We've discussed this a lot and tried to think of clever solutions but currently it's not possible as far as we are aware.

> BTW, I am curious as to why TAILWIND_MODE=watch prevents this issue. Does write the CSS file to a different location?

`TAILWIND_MODE=watch` does not use your build tools native dependency tracking, instead we start a separate process that watches the globs independently using chokidar. We register a single file as a dependency with webpack (it used to be in `~/.tailwindcss/touch/{randomstring}`) but now that webpack supports directory dependencies in PostCSS, we write just to a boring old temp file somewhere in your OS temp directory.

Whenever one of your template files that we are watching changes, we `touch` the "touch file" which imperatively triggers a rebuild. This has a lot of problems, like being prone to race conditions, making it appear like your build process is hanging, etc. which is why we have pushed to hard to get build tools updated to support the `dir-dependency` API so we can use native dependency tracking instead of running a separate dependency tracking system in parallel.

It has nothing to do with the CSS file output location, and only works because with that strategy we can watch a glob instead of being forced to only watch a directory (since that's all webpack supports).

In v2.2, the `TAILWIND_MODE=watch` option _will not work_ with webpack 5 if you have the persistent file cache enabled, as that was the reason we had to use that awful `~/.tailwindcss` hack and pollute people's hard drives. Now that webpack 5 supports the more reliable dependency tracking system, we use regular temp files to avoid writing crap to that folder.

---

Short of it is that for the best experience, you want your CSS file to be written to a different folder than your source files so you don't have to watch your CSS output directory for changes. We don't expect webpack to add glob support any time soon, so your best options are to be more specific in your `purge` config and rearrange your directories a little bit if necessary, or don't use webpack and instead use our CLI tool which supports globs properly.

**Comment by tleish** [comment]
I can certainly tell it's complicated.

One technique used in ruby gem [rubocop](https://github.com/rubocop/rubocop) was massaging GLOB inputs (I was a contributor).  That is, to convert certain GLOBs patterns to better matching GLOB patterns.  You could do something similar to this in TailwindCSS.  For example:

- `{blog,admin}/app/views/**/*.erb`
becomes
- `blog/app/views/`
- `admin/app/views/`

or 

- `../*/app/views/**/*.erb`

becomes

- `../dir1/app/views/`
- `../dir2/app/views/`
- `../dir3/app/views/`

After the directories are expanded, then pass them onto PostCSS.

FYI, I intend to use a more specific purge paths as the workaround, just trying provide suggestions to help you avoid dealing with guys like me so you can spend more time building cool stuff.

**Comment by tleish** [comment]
If it's helpful to others, I added something like the following to support dynamic directories without requiring updates to tailwind.config.js

```diff
+ const glob = require('glob');
+ const path = require('path');

module.exports = {
  mode: 'jit',
  future: {
    removeDeprecatedGapUtilities: true,
  },
  purge: {
    content: [
      './app/packs/**/*.{js,css,scss}',
-      '{blog,admin}/app/views/**/*.erb',
+      ...glob.sync("{blog,admin}/app/views").map((dir) => path.join(dir, '**/*.erb'))
-      '../*/app/views/**/*.erb',
+      ...glob.sync("../*/app/views").map((dir) => path.join(dir, '**/*.erb'))
    ],
    options: {
      whitelist: [],
    },
  },

```

**Comment by therealgilles** [comment]
I think the separation of build vs source directories is not too hard to understand. In my case, I have a ./build directory where generated files goes, but I also had './*.php' files and had added './tailwind*.js' to the purge patterns. It is really not obvious that this is going to turn into the top dir being added to the watch list (if I understand correctly).

**Comment by ihorvorotnov** [comment]
Just a note from a WordPress user - unfortunately, there's no easy way to separate source and build directory because of the way WordPress themes are structured:
```
your-theme-name/
  assets/
    css/
      main.css // output
  resources/
    pcss/
      main.pcss // input
  page.php // as you see, main theme templates are in the theme root directory (you can't move them)
  single.php
  front-page.php
  archive.php
  ...
```
So you have to register `'./**/*.php'` pattern for purging.

In my case, adding `TAILWIND_MODE=watch` helped resolve the looping issue, although it still builds twice at launch.

**Comment by adamwathan** [comment]
@ihorvorotnov Is that using Laravel Mix? I would recommend using our CLI tool, then globs will work as expected.

**Comment by ihorvorotnov** [comment]
@adamwathan yes, this is with latest Laravel Mix. I'm looking into splitting my workflow so Mix will still build everything for production but during development I'll be using Tailwind CLI.

**Comment by gerhardsletten** [comment]
For a setup in a wordpress theme I fixed this issue using glob to create a list of all php files to the purge option:

```js
const glob = require('glob')

module.exports = {
  mode: 'jit',
  purge: [
    // './**/*.php',
    './assets/**/*.js',
  ].concat(glob.sync("./**/*.php")),
```
